File: IDEA/src/ksp/kos/ideaplugin/psi/impl/KerboScriptScopeImpl.java
package ksp.kos.ideaplugin.psi.impl;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import ksp.kos.ideaplugin.psi.KerboScriptNamedElement;
import ksp.kos.ideaplugin.psi.KerboScriptScope;
import ksp.kos.ideaplugin.reference.Cache;
import ksp.kos.ideaplugin.reference.LocalScope;
import org.jetbrains.annotations.NotNull;

/**
 * Created on 07/01/16.
 *
 * @author ptasha
 */
public class KerboScriptScopeImpl extends KerboScriptElementImpl implements KerboScriptScope {
    private final Cache<LocalScope> cache = new Cache<>(this, new LocalScope());

    public KerboScriptScopeImpl(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public void register(KerboScriptNamedElement element) {
        getLocalScope().register(element);
    }

    @Override
    public PsiElement resolveFunction(KerboScriptNamedElement element) {
        KerboScriptNamedElement function = getLocalScope().resolveFunction(element);
        if (function==null) {
            return getScope().resolveFunction(element);
        }
        return function;
    }

    @Override
    public PsiElement resolveVariable(KerboScriptNamedElement element) {
        KerboScriptNamedElement variable = getLocalScope().resolveVariable(element);
        if (variable==null) {
            return getScope().resolveVariable(element);
        }
        return variable;
    }

    @Override
    public void clearCache() {
        getLocalScope().clear();
    }

    @NotNull
    private LocalScope getLocalScope() {
        return cache.getScope();
    }
}

File: IDEA/src/main/java/ksp/kos/ideaplugin/psi/impl/KerboScriptScopeImpl.kt
package ksp.kos.ideaplugin.psi.impl

import com.intellij.extapi.psi.ASTWrapperPsiElement
import com.intellij.lang.ASTNode
import ksp.kos.ideaplugin.psi.KerboScriptScope
import ksp.kos.ideaplugin.reference.Cache
import ksp.kos.ideaplugin.reference.context.LocalContext

/**
 * Created on 07/01/16.
 *
 * @author ptasha
 */
open class KerboScriptScopeImpl(node: ASTNode) : ASTWrapperPsiElement(node), KerboScriptScope {
    private var cache: Cache<LocalContext>? = null

    @Synchronized
    override fun getCachedScope(): LocalContext {
        val resolvedCache = cache ?: run {
            // It's possible that this triggers a re-walk of the entire file, which will itself construct a
            // LocalContext and cache for this scope. If we now overwrite that it'll be incorrect. Make sure to
            // check again if our cache is still null before charging ahead.
            val parentScope = this.scope.cachedScope
            cache ?: run {
                val newCache = Cache(this, LocalContext(parentScope))
                cache = newCache
                newCache
            }
        }
        return resolvedCache.scope
    }
}
